File: Packages/TeamTasksKit/Sources/TeamTasksCore/Logic/ReminderLeadTime.swift
import Foundation

/// How long before a task's due date its reminder fires (Réglages › Rappels). Persisted in `KeyValueStore`.
public enum ReminderLeadTime: String, Sendable, Hashable, Codable, CaseIterable, Identifiable {
    case atDueTime
    case fifteenMinutes
    case oneHour
    case oneDay
    /// No due-date reminders.
    case off

    /// Default when nothing is stored: 1 hour before.
    public static let `default`: ReminderLeadTime = .oneHour
    /// `KeyValueStore` key.
    public static let storageKey = "settings.reminderLeadTime"

    public var id: String { rawValue }

    /// French label for the settings picker.
    public var label: String {
        switch self {
        case .atDueTime: "À l'heure de l'échéance"
        case .fifteenMinutes: "15 minutes avant"
        case .oneHour: "1 heure avant"
        case .oneDay: "1 jour avant"
        case .off: "Aucun rappel"
        }
    }

    public var isEnabled: Bool { self != .off }

    /// When the reminder of a task due at `dueAt` fires, or nil when reminders are off.
    /// "1 jour avant" is one calendar day earlier at the same wall-clock time (23 or 25 hours across a
    /// daylight-saving transition), per `calendar`'s time zone.
    public func fireDate(forDueAt dueAt: Date, calendar: Calendar) -> Date? {
        switch self {
        case .atDueTime: dueAt
        case .fifteenMinutes: dueAt.addingTimeInterval(-15 * 60)
        case .oneHour: dueAt.addingTimeInterval(-60 * 60)
        case .oneDay: calendar.date(byAdding: .day, value: -1, to: dueAt) ?? dueAt.addingTimeInterval(-86_400)
        case .off: nil
        }
    }

    /// Stored value, or `.default` when missing or unreadable.
    public static func load(from store: any KeyValueStore) -> ReminderLeadTime {
        store.value(ReminderLeadTime.self, forKey: storageKey) ?? .default
    }

    public func save(to store: any KeyValueStore) {
        store.setValue(self, forKey: Self.storageKey)
    }
}

File: android/core/src/main/kotlin/io/github/notkanaa/equipe/core/logic/ReminderLeadTime.kt
package io.github.notkanaa.equipe.core.logic

import io.github.notkanaa.equipe.core.AppCalendar
import io.github.notkanaa.equipe.core.KeyValueStore
import io.github.notkanaa.equipe.core.setValue
import io.github.notkanaa.equipe.core.value
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Duration
import java.time.Instant

// Port of TeamTasksCore/Logic/ReminderLeadTime.swift.

/**
 * How long before a task's due date its reminder fires (Réglages › Rappels). Persisted in [KeyValueStore] under
 * [STORAGE_KEY] as a JSON string of [rawValue] (e.g. `"oneHour"`), like the iOS app.
 */
@Serializable
enum class ReminderLeadTime(val rawValue: String, val label: String) {
    @SerialName("atDueTime")
    AT_DUE_TIME("atDueTime", "À l’heure de l’échéance"),

    @SerialName("fifteenMinutes")
    FIFTEEN_MINUTES("fifteenMinutes", "15 minutes avant"),

    @SerialName("oneHour")
    ONE_HOUR("oneHour", "1 heure avant"),

    @SerialName("oneDay")
    ONE_DAY("oneDay", "1 jour avant"),

    /** No due-date reminders. */
    @SerialName("off")
    OFF("off", "Aucun rappel"),
    ;

    val id: String get() = rawValue

    val isEnabled: Boolean get() = this != OFF

    /**
     * When the reminder of a task due at [dueAt] fires, or null when reminders are off.
     * "1 jour avant" is one calendar day earlier at the same wall-clock time (23 or 25 hours across a daylight-saving
     * transition), per [calendar]'s time zone.
     */
    fun fireDate(dueAt: Instant, calendar: AppCalendar): Instant? = when (this) {
        AT_DUE_TIME -> dueAt
        FIFTEEN_MINUTES -> dueAt.minus(Duration.ofMinutes(15))
        ONE_HOUR -> dueAt.minus(Duration.ofHours(1))
        ONE_DAY -> dueAt.atZone(calendar.zone).minusDays(1).toInstant()
        OFF -> null
    }

    fun save(store: KeyValueStore) {
        store.setValue(STORAGE_KEY, this)
    }

    companion object {
        /** Default when nothing is stored: 1 hour before. */
        val DEFAULT: ReminderLeadTime = ONE_HOUR

        /** [KeyValueStore] key. */
        const val STORAGE_KEY: String = "settings.reminderLeadTime"

        /** The lead time whose raw value is [rawValue], or null. */
        fun fromRawValue(rawValue: String): ReminderLeadTime? = entries.firstOrNull { it.rawValue == rawValue }

        /** Stored value, or [DEFAULT] when missing or unreadable. */
        fun load(store: KeyValueStore): ReminderLeadTime = store.value<ReminderLeadTime>(STORAGE_KEY) ?: DEFAULT
    }
}
